### PR TITLE
RDK-35406 : XCAST unregisterApplications API

### PR DESCRIPTION
**Pull request by acheri988**
Reason for change:
XCAST unregisterApplications API
Test Procedure: None
Risks: Low

Change-Id: I205386a77bf5a20a08ba6012265e975d1804f2c8 Signed-off-by:Anooj Cheriyan <Anooj_Cheriyan@comcast.com>

### CODE REVIEW

File: XCast/XCast.cpp [hunk 1]
```diff
@@ -96,12 +97,11 @@ bool XCast::m_xcastEnable= false;
...
 bool XCast::m_standbyBehavior = false;
 bool XCast::m_enableStatus = false;
-string strDyAppConfig = "";
 
 IARM_Bus_PWRMgr_PowerState_t XCast::m_powerState = IARM_BUS_PWRMGR_POWERSTATE_STANDBY;
 
 XCast::XCast() : PluginHost::JSONRPC()
-, m_apiVersionNumber(1)
+, m_apiVersionNumber(1), m_isDynamicRegistrationsToDo(false)
 {
```
Comment (apatel859) [inline XCast/XCast.cpp:105]: ToDo -> Pending

File: XCast/XCast.cpp [hunk 2]
```diff
@@ -96,12 +97,11 @@ bool XCast::m_xcastEnable= false;
 string XCast::m_friendlyName = "";
 bool XCast::m_standbyBehavior = false;
 bool XCast::m_enableStatus = false;
-string strDyAppConfig = "";
 
 IARM_Bus_PWRMgr_PowerState_t XCast::m_powerState = IARM_BUS_PWRMGR_POWERSTATE_STANDBY;
 
 XCast::XCast() : PluginHost::JSONRPC()
-, m_apiVersionNumber(1)
+, m_apiVersionNumber(1), m_isDynamicRegistrationsToDo(false)
```
Comment (apatel859) [inline XCast/XCast.cpp:104]: ToDo -> Required

File: XCast/XCast.cpp [hunk 3]
```diff
@@ -356,6 +357,325 @@ uint32_t XCast::getProtocolVersion(const JsonObject& parameters, JsonObject& res
     returnResponse(true);
 }
 
+bool XCast::getEntryFromAppLaunchParamList (const char* appName, DynamicAppConfig* reqParam){
```
Comment (apatel859) [inline XCast/XCast.cpp:360]: retAppConfig

File: XCast/XCast.cpp [hunk 4]
```diff
@@ -356,6 +357,325 @@ uint32_t XCast::getProtocolVersion(const JsonObject& parameters, JsonObject& res
...
+            if (0 == strcmp (regAppLaunchParam->appName, appName)) {
+                isEntryFound = true;
+                int iNameLen = strlen (regAppLaunchParam->appName);
+                reqParam->appName = (char*) malloc (iNameLen+1);
+                memset (reqParam->appName, '\0', iNameLen+1);
+                strcpy (reqParam->appName, regAppLaunchParam->appName);
+
+                if (regAppLaunchParam->query) {
+                    int iQueryLen = strlen (regAppLaunchParam->query);
+                    reqParam->query = (char*) malloc (iQueryLen+1);
```
Comment (apatel859) [inline XCast/XCast.cpp:373]: think of replacing malloc with static array

File: XCast/XCast.cpp [hunk 5]
```diff
@@ -356,6 +357,325 @@ uint32_t XCast::getProtocolVersion(const JsonObject& parameters, JsonObject& res
...
+                    entriesTodelete.push_back(index);
+                    found = true;
+                    break;
+                }
+                index ++;
+            }
+            if (!found) {
+                LOGINFO("%s not existing in the dynamic cache", appNameToDelete.c_str());
+            }
+            ret = (ret&&found);
```
Comment (apatel859) [inline XCast/XCast.cpp:427]: chcck ret if A is found and D is not found

File: XCast/XCast.cpp [hunk 6]
```diff
@@ -356,6 +357,325 @@ uint32_t XCast::getProtocolVersion(const JsonObject& parameters, JsonObject& res
...
+        deleteFromDynamicAppCache (appsToDelete);
+
+        LOGINFO("appConfigList count: %d", appConfigList.size());
+        //Update the new entries here.
+        {lock_guard<mutex> lck(m_appConfigMutex);
+            for (DynamicAppConfig* pDynamicAppConfig : appConfigList) {
+                m_appConfigCache.push_back(pDynamicAppConfig);
+            }
+            //If empty list is passed cache need to be cleared
+            if(!appConfigList.size()){
```
Comment (apatel859) [inline XCast/XCast.cpp:665]: move this out as if appConfigList is 0 u need to clear anyway

File: XCast/XCast.h [hunk 7]
```diff
@@ -103,6 +113,12 @@ class XCast : public PluginHost::IPlugin, public PluginHost::JSONRPC, public RtN
...
     void onLocateCastTimer();
     void getUrlFromAppLaunchParams (const char *app_name, const char *payload, const char *query_string, const char *additional_data_url, char *url);
+    bool getEntryFromAppLaunchParamList (const char* appName, DynamicAppConfig* reqParam);
+    void dumpDynamicAppConfigCache(string strListName, std::vector<DynamicAppConfig*> appConfigList);
+    bool deleteFromDynamicAppCache(string strAppNames);
+    bool deleteFromDynamicAppCache(vector<string>& appsToDelete);
+    void updateDynamicAppCache(string strApps);
+
     /**
      * Check whether the xdial service is allowed in this device.
```
Comment (apatel859) [inline XCast/XCast.h:123]: update API documentation for both register and unregister with complete detials

**Comment by acheri988** [comment]
New PR created: https://github.com/rdkcentral/rdkservices/pull/3182

**Pull request closed without merging**